unison-src/transcripts/doc-formatting.output.md

```diff
@@ -0,0 +1,469 @@
+
+```unison
+escaping = [: Docs look [: like \@this \:] :]
+```
+
+```ucm
+
+  I found and typechecked these definitions in scratch.u. If you
+  do an `add` or `update`, here's how your codebase would
+  change:
+  
+    ⍟ These new definitions are ok to `add`:
+    
+      escaping : Doc
+   
+  Now evaluating any watch expressions (lines starting with
+  `>`)... Ctrl+C cancels.
+
+```
+```ucm
+.> view escaping
+
+  escaping : Doc
+  escaping = [: Docs look [: like \@this \:] :]
+
+```
+(Alas you can't have `\@` or `\:]` in your doc, as there's currently no way to 'unescape' them.)
+
+```unison
+-- Note that -- comments are preserved within doc literals.
+commented = [:
+  example:
+
+    -- a comment
+    f x = x + 1
+:]
+```
+
+```ucm
+
+  I found and typechecked these definitions in scratch.u. If you
+  do an `add` or `update`, here's how your codebase would
+  change:
+  
+    ⍟ These new definitions are ok to `add`:
+    
+      commented : Doc
+   
+  Now evaluating any watch expressions (lines starting with
+  `>`)... Ctrl+C cancels.
+
+```
+```ucm
+.> view commented
+
+  commented : Doc
+  commented =
+    [: 
+    example:
+    
+      -- a comment
+      f x = x + 1
+     :]
+
+```
+### Indenting, and paragraph reflow
+
+Handling of indenting in docs between the parser and pretty-printer is a bit fiddly.
+
+```unison
+-- The leading and trailing spaces are stripped from the stored Doc by the
+-- lexer, and one leading and trailing space is inserted again on view/edit
+-- by the pretty-printer.
+doc1 = [:   hi   :]
+```
+
+```ucm
+
+  I found and typechecked these definitions in scratch.u. If you
+  do an `add` or `update`, here's how your codebase would
+  change:
+  
+    ⍟ These new definitions are ok to `add`:
+    
+      doc1 : Doc
+   
+  Now evaluating any watch expressions (lines starting with
+  `>`)... Ctrl+C cancels.
+
+```
+```ucm
+.> view doc1
+
+  doc1 : Doc
+  doc1 = [: hi :]
+
+```
+```unison
+-- Lines (apart from the first line, i.e. the bit between the [: and the
+-- first newline) are unindented until at least one of
+-- them hits the left margin (by a post-processing step in the parser).
+-- You may not notice this because the pretty-printer indents them again on
+-- view/edit.
+doc2 = [: hello
+            - foo
+            - bar
+          and the rest. :]
+```
+
+```ucm
+
+  I found and typechecked these definitions in scratch.u. If you
+  do an `add` or `update`, here's how your codebase would
+  change:
+  
+    ⍟ These new definitions are ok to `add`:
+    
+      doc2 : Doc
+   
+  Now evaluating any watch expressions (lines starting with
+  `>`)... Ctrl+C cancels.
+
+```
+```ucm
+.> view doc2
+
+  doc2 : Doc
+  doc2 =
+    [: hello
+      - foo
+      - bar
+    and the rest. :]
+
+```
+```unison
+doc3 = [: When Unison identifies a paragraph, it removes any newlines from it before storing it, and then reflows the paragraph text to fit the display window on display/view/edit.
```

Comment:
@atacratic wow this reflowing is great!! I had wanted to do that but found it was more complicated than I was willing to take on at the time so it's cool that you figured it out.